### PR TITLE
AirfRANS: vol-weight=5.0 & 7.0 + EMA=0.999 — filling 3x→10x gap

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

vol-weight=3x (PR #3127) gave a strong surface improvement (-17%, surface_mse=0.000381!) but the volume improvement was modest. vol-weight=10x was too aggressive and hurt both metrics. The sweet spot for volume MSE improvement likely lies between 3x and 10x. We test 5x and 7x to find it, pairing with EMA=0.999 which proved essential for surface quality.

**Target:** Keep surface_mse competitive (ideally near 0.000381) while pushing vol_mse toward the SpiderSolver target of ~0.0017.

## Instructions

Run **two trials** using the mandatory AF config (AdamW lr=6e-4, T_max=50, gc=1.0, WD=1e-2, EMA=0.999, 2L/256d/8H). Use `--wandb_group af-vol-weight-5x-7x` to group runs.

**Trial 1 — vol-weight=5.0:**
```bash
cd target/icml2026 && python train.py --dataset airfrans --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 8 --epochs 999 --ema-decay 0.999 --vol-loss-weight 5.0 --wandb_group af-vol-weight-5x-7x
```

**Trial 2 — vol-weight=7.0:**
```bash
cd target/icml2026 && python train.py --dataset airfrans --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 8 --epochs 999 --ema-decay 0.999 --vol-loss-weight 7.0 --wandb_group af-vol-weight-5x-7x
```

**What to report:** For each trial, report `val_primary/surface_mse` and `full_val/volume_mse` at the best validation checkpoint. Compare both against baseline. Flag whether vol_mse is trending toward 0.0017.

## Baseline

Current best — PR #3050 (stark — EMA=0.999 + T_max=50, 2L/256d, AdamW lr=6e-4, gc=1.0, WD=1e-2):
- **val_primary/surface_mse:** 0.000459 (target to beat)
- **full_val/volume_mse:** 0.002777 (target to beat; SpiderSolver target ~0.0017)
- **W&B run:** z6pry4b9
- **Reproduce:** `cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --ema-decay 0.999`

## Context: vol-weight sweep so far

| vol-weight | surface_mse | vol_mse | vs baseline surface | vs baseline vol |
|---|---|---|---|---|
| 1.0 (baseline) | 0.000459 | 0.002777 | — | — |
| 1.5 (tanjiro, WIP) | TBD | TBD | — | — |
| 2.0 (violet, WIP) | TBD | TBD | — | — |
| 3.0 (PR #3127) | 0.000381 | ~0.002777 | **-17%** | ~flat |
| **5.0 (this PR, Trial 1)** | TBD | TBD | — | — |
| **7.0 (this PR, Trial 2)** | TBD | TBD | — | — |
| 10.0 (nezuko, WIP) | worse | worse | — | — |
| 30.0 (megumi, WIP) | much worse | TBD | — | — |

Dead ends: 3L depth, EMA<0.999, T_max≠50, vol-weight=10x

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT vegeta:

## Results

### Trial 1 — vol-weight=5.0 (W&B run: dxna9k23)

**At best validation checkpoint (epoch 640):**
| Metric | Value | Baseline | Change |
|---|---|---|---|
| val_primary/surface_mse | **0.000432** | 0.000459 | **-5.9%** |
| full_val/volume_mse | **0.003094** | 0.002777 | +11.4% |
| test surface_mse | 0.001224 | — | — |
| test volume_mse | 0.003941 | — | — |

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py --dataset airfrans --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 8 --epochs 999 --ema-decay 0.999 --vol-loss-weight 5.0 --wandb-group af-vol-weight-5x-7x --agent vegeta --wandb-name "vegeta/af-vol-weight-5x-ema0999"
```

### Trial 2 — vol-weight=7.0 (W&B run: 20nl7euu)

**At best validation checkpoint (epoch 322):**
| Metric | Value | Baseline | Change |
|---|---|---|---|
| val_primary/surface_mse | **0.000438** | 0.000459 | **-4.6%** |
| full_val/volume_mse | **0.005187** | 0.002777 | +86.8% |
| test surface_mse | 0.002579 | — | — |
| test volume_mse | 0.006228 | — | — |

**Command:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 python train.py --dataset airfrans --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 8 --epochs 999 --ema-decay 0.999 --vol-loss-weight 7.0 --wandb-group af-vol-weight-5x-7x --agent vegeta --wandb-name "vegeta/af-vol-weight-7x-ema0999"
```

### What happened

Both trials beat the baseline on surface_mse, but with an important caveat:

**Surface quality:** Both vol-weight=5.0 and vol-weight=7.0 improved surface_mse vs baseline (-5.9% and -4.6% respectively). This is consistent with the vol-weight=3.0 result (-17%), suggesting that adding more volume emphasis can still help surface predictions through better feature learning.

**Volume quality:** Neither trial improved volume_mse vs baseline. Vol-weight=5.0 was +11.4% worse, and vol-weight=7.0 was +86.8% worse. This is counterintuitive — increasing the volume loss weight degraded volume performance. The model seems to overfit to volume training patterns at higher weights, hurting generalization on the volume validation set.

**8H vs 4H:** Note these trials used 8 attention heads (as specified in the PR instructions) vs the baseline's 4 heads, which makes the comparison imperfect. The head count change may be responsible for some of the observed differences.

**Training dynamics:** Both runs benefited greatly from extended training past T_max=50 (cosine warm restarts). Trial 1 found its best checkpoint at epoch 640, and Trial 2 at epoch 322. The surface metric continued improving well past the first cosine cycle.

**Vol_mse vs SpiderSolver target (~0.0017):** Neither trial moved toward the SpiderSolver target. Vol-weight=5.0 gave vol_mse=0.003094 and vol-weight=7.0 gave vol_mse=0.005187, both worse than the baseline's 0.002777.

### Bug fix

Fixed a `primary_metric_key` shadowing bug in train.py (line 1639) — a local variable assignment shadowed the function of the same name, causing an `UnboundLocalError` at startup. Also added `vol_loss_weight` as a new configurable parameter to `TrainConfig` and wired it through all three loss functions.

### Suggested follow-ups

1. **Re-run with 4H instead of 8H** to make the comparison fair against the baseline.
2. **vol-weight=3.0 remains the sweet spot** based on the sweep so far — it gave -17% surface improvement with flat volume. Values of 5x and 7x degrade volume.
3. **Investigate why higher vol-weight hurts volume generalization** — possible that the increased gradient magnitude from volume loss causes optimization instability or overfitting.
4. **Consider asymmetric weighting schedules** — start with high vol-weight to learn volume features, then anneal toward 1.0 for balanced fine-tuning.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3166 because ADVISOR: vol-weight=5x/7x both worse than 10x champion on BOTH metrics. 5x: surface=0.000432 (+46%), vol=0.003094 (+52%) (W&B: dxna9k23). 7x: surface=0.000438 (+48%), vol=0.005187 (+154%), then diverged (W&B: 20nl7euu). Also used 8H instead of champion's 4H. vol-weight sweep now complete: 2x, 5x, 7x all worse, 10x champion, 15x/20x/30x pending/dead. 10x+EMA=0.999 confirmed as AF sweet spot. NEW DIRECTIVE: All AF work now targets volume closure only..
